File: shipley_checks/src/encyclopedia/sections/s3_maintenance_profile.R
# ==============================================================================
# SECTION 3: MAINTENANCE PROFILE (LABOR REQUIREMENTS)
# ==============================================================================
#
# PURPOSE:
#   Translate CSR strategy and plant traits into practical maintenance advice
#   including growth rate, pruning needs, seasonal tasks, and time commitment.
#
# DATA SOURCES:
#   - C, S, R: CSR strategy scores (0-1 scale, 99.88% coverage)
#   - try_growth_form: Growth form classification
#   - height_m: Maximum height in meters
#   - try_leaf_phenology: Deciduous vs evergreen
#   - decomposition_rating: Decomposition rate (0-10 scale if available)
#   - decomposition_confidence: Confidence in decomposition rating
#
# OUTPUT FORMAT:
#   Markdown text with maintenance level header and 4 subsections:
#   🌿 Growth Rate | 🍂 Seasonal Tasks | ♻️ Waste Management | ⏰ Time Commitment
#
# DEPENDENCIES:
#   - utils/categorization.R: get_csr_category()
#
# ==============================================================================

library(dplyr)
library(glue)

# Source dependencies
if (!exists("get_csr_category")) {
  source("shipley_checks/src/encyclopedia/utils/categorization.R")
}

# ==============================================================================
# MAIN GENERATION FUNCTION
# ==============================================================================

#' Generate Section 3: Maintenance Profile
#'
#' Creates maintenance and labor requirement advice based on CSR strategy
#' and plant traits.
#'
#' @param plant_row Single-row data frame with plant data
#' @return Character string of markdown-formatted maintenance profile
#'
#' @details
#' ALGORITHM:
#'   1. Determine overall maintenance level from CSR strategy
#'   2. Calculate growth rate and pruning needs
#'   3. Generate seasonal task recommendations
#'   4. Assess waste management (decomposition characteristics)
#'   5. Estimate annual time commitment
#'
#' @export
generate_section_3_maintenance_profile <- function(plant_row) {

  # ===========================================================================
  # STEP 1: Extract CSR strategy
  # ===========================================================================

  csr_C <- plant_row$C
  csr_S <- plant_row$S
  csr_R <- plant_row$R

  csr_category <- get_csr_category(csr_C, csr_S, csr_R)

  # ===========================================================================
  # STEP 2: Determine overall maintenance level
  # ===========================================================================
  # RATIONALE: C-dominant = high maintenance (fast growth, vigorous)
  #            S-dominant = low maintenance (slow growth, hardy)
  #            R-dominant = medium (opportunistic but not as vigorous as C)

  maintenance_level <- calculate_maintenance_level(csr_C, csr_S, csr_R)

  # ===========================================================================
  # STEP 3: Growth rate section
  # ===========================================================================

  growth_section <- generate_growth_rate_advice(csr_C, csr_S, csr_R, csr_category,
                                                  plant_row$try_growth_form,
                                                  plant_row$height_m)

  # ===========================================================================
  # STEP 4: Seasonal tasks section
  # ===========================================================================

  seasonal_section <- generate_seasonal_tasks(plant_row$try_leaf_phenology,
                                                plant_row$try_growth_form,
                                                csr_C, csr_S)

  # ===========================================================================
  # STEP 5: Waste management section
  # ===========================================================================

  waste_section <- generate_waste_management(plant_row$decomposition_rating,
                                               plant_row$decomposition_confidence,
                                               csr_S,
                                               plant_row$try_leaf_phenology)

  # ===========================================================================
  # STEP 6: Time commitment estimate
  # ===========================================================================

  time_section <- estimate_time_commitment(maintenance_level,
                                             plant_row$try_growth_form,
                                             plant_row$try_leaf_phenology)

  # ===========================================================================
  # STEP 7: Assemble complete section
  # ===========================================================================

  maintenance_profile <- glue("
    ## Maintenance Profile

    **Maintenance Level: {toupper(maintenance_level)}**

    {growth_section}

    {seasonal_section}

    {waste_section}

    {time_section}
  ")

  return(as.character(maintenance_profile))
}

# ==============================================================================
# HELPER FUNCTIONS
# ==============================================================================

#' Calculate Overall Maintenance Level
#' @keywords internal
calculate_maintenance_level <- function(csr_C, csr_S, csr_R) {

  if (any(is.na(c(csr_C, csr_S, csr_R)))) {
    return("MEDIUM")
  }

  # Normalize to 0-1 scale if in percentage form (0-100)
  if (csr_C > 1 || csr_S > 1 || csr_R > 1) {
    csr_C <- csr_C / 100
    csr_S <- csr_S / 100
    csr_R <- csr_R / 100
  }

  # RATIONALE: Thresholds based on CSR triangle partitioning
  if (csr_C > 0.6) {
    return("HIGH")
  } else if (csr_S > 0.6) {
    return("LOW")
  } else if (csr_R > 0.6) {
    return("MEDIUM")
  } else {
    # Mixed strategies
    if (csr_C > csr_S && csr_C > csr_R) {
      return("MEDIUM-HIGH")
    } else if (csr_S > csr_C && csr_S > csr_R) {
      return("LOW-MEDIUM")
    } else {
      return("MEDIUM")
    }
  }
}

#' Generate Growth Rate Advice
#' @keywords internal
generate_growth_rate_advice <- function(csr_C, csr_S, csr_R, csr_category,
                                         growth_form, height_m) {

  if (any(is.na(c(csr_C, csr_S, csr_R)))) {
    return("🌿 **Growth Rate**: Unknown")
  }

  # Normalize to 0-1 scale if in percentage form (0-100)
  if (csr_C > 1 || csr_S > 1 || csr_R > 1) {
    csr_C <- csr_C / 100
    csr_S <- csr_S / 100
    csr_R <- csr_R / 100
  }

  # Determine growth rate
  if (csr_C > 0.6) {
    growth_rate <- "Fast"
    growth_desc <- "Vigorous grower, can outcompete neighbors"
  } else if (csr_S > 0.6) {
    growth_rate <- "Slow"
    growth_desc <- "Slow but steady growth, compact habit"
  } else if (csr_R > 0.6) {
    growth_rate <- "Rapid"
    growth_desc <- "Quick to establish, opportunistic growth"
  } else {
    growth_rate <- "Moderate"
    growth_desc <- "Steady growth rate"
  }

  # Pruning frequency
  pruning_advice <- ""
  if (!is.na(growth_form) && growth_form %in% c("tree", "shrub")) {
    if (csr_C > 0.6) {
      pruning_advice <- "→ Annual pruning recommended to control vigor"
    } else if (csr_S > 0.6) {
      pruning_advice <- "→ Minimal pruning needed (every 2-3 years)"
    } else {
      pruning_advice <- "→ Prune as needed to maintain shape"
    }
  } else if (!is.na(growth_form) && growth_form == "climber") {
    if (csr_C > 0.6) {
      pruning_advice <- "→ Requires regular training and pruning"
    } else {
      pruning_advice <- "→ Light pruning to guide growth"
    }
  } else {
    # Herbaceous plants
    if (csr_C > 0.6) {
      pruning_advice <- "→ May require cutting back to prevent spreading"
    } else {
      pruning_advice <- "→ Cut back after flowering if desired"
    }
  }

  growth_section <- glue("
    🌿 **Growth Rate**: {growth_rate} ({csr_category} strategy)
       → {growth_desc}
       {pruning_advice}
  ")

  return(as.character(growth_section))
}

#' Generate Seasonal Tasks Advice
#' @keywords internal
generate_seasonal_tasks <- function(leaf_phenology, growth_form, csr_C, csr_S) {

  # Normalize CSR if in percentage form
  if (!is.na(csr_C) && csr_C > 1) csr_C <- csr_C / 100
  if (!is.na(csr_S) && csr_S > 1) csr_S <- csr_S / 100

  tasks <- c()

  # Spring tasks
  if (!is.na(growth_form) && growth_form %in% c("tree", "shrub")) {
    spring_task <- "Spring: Shape after frost risk passes"
  } else {
    spring_task <- "Spring: Remove dead growth, apply mulch"
  }
  tasks <- c(tasks, spring_task)

  # Summer tasks
  if (!is.na(csr_C) && csr_C > 0.6) {
    summer_task <- "Summer: Monitor for excessive growth, deadhead spent flowers"
  } else {
    summer_task <- "Summer: Minimal intervention, occasional deadheading"
  }
  tasks <- c(tasks, summer_task)

  # Autumn tasks
  if (!is.na(leaf_phenology) && leaf_phenology == "deciduous") {
    autumn_task <- "Autumn: Rake fallen leaves for compost"
  } else {
    autumn_task <- "Autumn: Minimal leaf cleanup (evergreen foliage)"
  }
  tasks <- c(tasks, autumn_task)

  # Winter tasks
  if (!is.na(growth_form) && growth_form %in% c("tree", "shrub")) {
    winter_task <- "Winter: Structural pruning while dormant (if deciduous)"
  } else {
    winter_task <- "Winter: Protect from frost if tender"
  }
  tasks <- c(tasks, winter_task)

  seasonal_section <- glue("
    🍂 **Seasonal Tasks**:
       → {paste(tasks, collapse = '\n       → ')}
  ")

  return(as.character(seasonal_section))
}

#' Generate Waste Management Advice
#' @keywords internal
generate_waste_management <- function(decomp_rating, decomp_confidence,
                                       csr_S, leaf_phenology) {

  # Convert to numeric if needed
  decomp_rating <- suppressWarnings(as.numeric(decomp_rating))
  decomp_confidence <- suppressWarnings(as.numeric(decomp_confidence))

  # Normalize CSR-S if in percentage form
  if (!is.na(csr_S) && csr_S > 1) csr_S <- csr_S / 100

  # Use decomposition rating if available and confident
  if (!is.na(decomp_rating) && !is.na(decomp_confidence) && decomp_confidence >= 0.5) {

    if (decomp_rating >= 7) {
      decomp_desc <- "Fast-decomposing foliage"
      compost_advice <- "Excellent for compost; breaks down quickly"
      mulch_advice <- "Good green mulch but short-lived"
    } else if (decomp_rating >= 4) {
      decomp_desc <- "Moderate decomposition rate"
      compost_advice <- "Suitable for compost; mix with other materials"
      mulch_advice <- "Decent mulch material"
    } else {
      decomp_desc <- "Slow-decomposing foliage"
      compost_advice <- "Add to compost in thin layers; may take time"
      mulch_advice <- "Excellent for long-lasting mulch"
    }

  } else {
    # Fall back to CSR strategy
    if (!is.na(csr_S) && csr_S > 0.6) {
      decomp_desc <- "Likely slow-decomposing (stress-tolerator strategy)"
      compost_advice <- "Tough foliage; shred before composting"
      mulch_advice <- "Good for long-lasting mulch"
    } else {
      decomp_desc <- "Moderate decomposition rate"
      compost_advice <- "Suitable for composting"
      mulch_advice <- "Can be used as mulch"
    }
  }

  # Add note about volume if deciduous
  if (!is.na(leaf_phenology) && leaf_phenology == "deciduous") {
    volume_note <- "Seasonal leaf drop creates moderate waste volume"
  } else {
    volume_note <- "Minimal waste volume (evergreen)"
  }

  waste_section <- glue("
    ♻️ **Waste Management**:
       → {decomp_desc}
       → {compost_advice}
       → {mulch_advice}
       → {volume_note}
  ")

  return(as.character(waste_section))
}

#' Estimate Annual Time Commitment
#' @keywords internal
estimate_time_commitment <- function(maintenance_level, growth_form, leaf_phenology) {

  # Base time by maintenance level
  base_minutes <- switch(maintenance_level,
    "LOW" = 15,
    "LOW-MEDIUM" = 30,
    "MEDIUM" = 60,
    "MEDIUM-HIGH" = 120,
    "HIGH" = 240,
    60  # default
  )

  # Adjust for growth form
  if (!is.na(growth_form)) {
    if (growth_form %in% c("tree", "shrub")) {
      base_minutes <- base_minutes * 1.5
    } else if (growth_form == "climber") {
      base_minutes <- base_minutes * 2
    }
  }

  # Adjust for deciduous (more cleanup)
  if (!is.na(leaf_phenology) && leaf_phenology == "deciduous") {
    base_minutes <- base_minutes * 1.3
  }

  # Format time
  if (base_minutes < 60) {
    time_text <- sprintf("~%d minutes per year", round(base_minutes))
  } else if (base_minutes < 180) {
    time_text <- sprintf("~%.1f hours per year", base_minutes / 60)
  } else {
    time_text <- sprintf("~%.0f hours per year", round(base_minutes / 60))
  }

  time_section <- glue("
    ⏰ **Time Commitment**: {time_text}
  ")

  return(as.character(time_section))
}

File: shipley_checks/src/Stage_4/guild_scorer_rust/src/encyclopedia/sections/s3_maintenance_profile.rs
//! Section 3: Maintenance Profile (Labor Requirements)
//!
//! Translates CSR strategy and plant traits into practical maintenance advice.
//!
//! Ported from R: shipley_checks/src/encyclopedia/sections/s3_maintenance_profile.R

use crate::encyclopedia::utils::categorization::{
    calculate_maintenance_level, get_csr_category, MaintenanceLevel, CsrCategory,
};

/// Plant data for maintenance profile
pub struct MaintenanceData {
    pub csr_c: Option<f64>,
    pub csr_s: Option<f64>,
    pub csr_r: Option<f64>,
    pub growth_form: Option<String>,
    pub height_m: Option<f64>,
    pub leaf_phenology: Option<String>,
    pub decomposition_rating: Option<f64>,
    pub decomposition_confidence: Option<f64>,
}

/// Generate Section 3: Maintenance Profile
pub fn generate_maintenance_profile(data: &MaintenanceData) -> String {
    let mut sections = Vec::new();

    sections.push("## Maintenance Profile".to_string());

    // Overall maintenance level
    let level = calculate_maintenance_level(
        data.csr_c.unwrap_or(f64::NAN),
        data.csr_s.unwrap_or(f64::NAN),
        data.csr_r.unwrap_or(f64::NAN),
    ).unwrap_or(MaintenanceLevel::Medium);

    sections.push(format!("\n**Maintenance Level: {}**", level.as_str()));

    // Growth rate
    sections.push(generate_growth_rate_advice(data));

    // Seasonal tasks
    sections.push(generate_seasonal_tasks(data));

    // Waste management
    sections.push(generate_waste_management(data));

    // Time commitment
    sections.push(estimate_time_commitment(&level, data));

    sections.join("\n\n")
}

fn generate_growth_rate_advice(data: &MaintenanceData) -> String {
    let (c, s, r) = (
        data.csr_c.unwrap_or(f64::NAN),
        data.csr_s.unwrap_or(f64::NAN),
        data.csr_r.unwrap_or(f64::NAN),
    );

    if c.is_nan() || s.is_nan() || r.is_nan() {
        return "🌿 **Growth Rate**: Unknown".to_string();
    }

    // Normalize
    let (c, s, r) = if c > 1.0 || s > 1.0 || r > 1.0 {
        (c / 100.0, s / 100.0, r / 100.0)
    } else {
        (c, s, r)
    };

    let csr_cat = get_csr_category(c, s, r).unwrap_or(CsrCategory::CSR);

    let (rate, desc) = if c > 0.6 {
        ("Fast", "Vigorous grower, can outcompete neighbors")
    } else if s > 0.6 {
        ("Slow", "Slow but steady growth, compact habit")
    } else if r > 0.6 {
        ("Rapid", "Quick to establish, opportunistic growth")
    } else {
        ("Moderate", "Steady growth rate")
    };

    let pruning = match data.growth_form.as_deref() {
        Some("tree") | Some("shrub") if c > 0.6 => "→ Annual pruning recommended to control vigor",
        Some("tree") | Some("shrub") if s > 0.6 => "→ Minimal pruning needed (every 2-3 years)",
        Some("tree") | Some("shrub") => "→ Prune as needed to maintain shape",
        Some("climber") if c > 0.6 => "→ Requires regular training and pruning",
        Some("climber") => "→ Light pruning to guide growth",
        _ if c > 0.6 => "→ May require cutting back to prevent spreading",
        _ => "→ Cut back after flowering if desired",
    };

    format!(
        "🌿 **Growth Rate**: {} ({} strategy)\n   → {}\n   {}",
        rate,
        csr_cat.as_str(),
        desc,
        pruning
    )
}

fn generate_seasonal_tasks(data: &MaintenanceData) -> String {
    let c = data.csr_c.map(|v| if v > 1.0 { v / 100.0 } else { v });

    let spring = match data.growth_form.as_deref() {
        Some("tree") | Some("shrub") => "Spring: Shape after frost risk passes",
        _ => "Spring: Remove dead growth, apply mulch",
    };

    let summer = match c {
        Some(c) if c > 0.6 => "Summer: Monitor for excessive growth, deadhead spent flowers",
        _ => "Summer: Minimal intervention, occasional deadheading",
    };

    let autumn = match data.leaf_phenology.as_deref() {
        Some("deciduous") => "Autumn: Rake fallen leaves for compost",
        _ => "Autumn: Minimal leaf cleanup (evergreen foliage)",
    };

    let winter = match data.growth_form.as_deref() {
        Some("tree") | Some("shrub") => "Winter: Structural pruning while dormant (if deciduous)",
        _ => "Winter: Protect from frost if tender",
    };

    format!(
        "🍂 **Seasonal Tasks**:\n   → {}\n   → {}\n   → {}\n   → {}",
        spring, summer, autumn, winter
    )
}

fn generate_waste_management(data: &MaintenanceData) -> String {
    let s = data.csr_s.map(|v| if v > 1.0 { v / 100.0 } else { v });

    let (decomp_desc, compost, mulch) = if let (Some(rating), Some(conf)) =
        (data.decomposition_rating, data.decomposition_confidence)
    {
        if conf >= 0.5 {
            if rating >= 7.0 {
                ("Fast-decomposing foliage",
                 "Excellent for compost; breaks down quickly",
                 "Good green mulch but short-lived")
            } else if rating >= 4.0 {
                ("Moderate decomposition rate",
                 "Suitable for compost; mix with other materials",
                 "Decent mulch material")
            } else {
                ("Slow-decomposing foliage",
                 "Add to compost in thin layers; may take time",
                 "Excellent for long-lasting mulch")
            }
        } else {
            get_decomp_from_csr(s)
        }
    } else {
        get_decomp_from_csr(s)
    };

    let volume = match data.leaf_phenology.as_deref() {
        Some("deciduous") => "Seasonal leaf drop creates moderate waste volume",
        _ => "Minimal waste volume (evergreen)",
    };

    format!(
        "♻️ **Waste Management**:\n   → {}\n   → {}\n   → {}\n   → {}",
        decomp_desc, compost, mulch, volume
    )
}

fn get_decomp_from_csr(csr_s: Option<f64>) -> (&'static str, &'static str, &'static str) {
    match csr_s {
        Some(s) if s > 0.6 => (
            "Likely slow-decomposing (stress-tolerator strategy)",
            "Tough foliage; shred before composting",
            "Good for long-lasting mulch"
        ),
        _ => (
            "Moderate decomposition rate",
            "Suitable for composting",
            "Can be used as mulch"
        ),
    }
}

fn estimate_time_commitment(level: &MaintenanceLevel, data: &MaintenanceData) -> String {
    let base_minutes: f64 = match level {
        MaintenanceLevel::Low => 15.0,
        MaintenanceLevel::LowMedium => 30.0,
        MaintenanceLevel::Medium => 60.0,
        MaintenanceLevel::MediumHigh => 120.0,
        MaintenanceLevel::High => 240.0,
    };

    let mut minutes = base_minutes;

    // Adjust for growth form
    match data.growth_form.as_deref() {
        Some("tree") | Some("shrub") => minutes *= 1.5,
        Some("climber") => minutes *= 2.0,
        _ => {}
    }

    // Adjust for deciduous
    if data.leaf_phenology.as_deref() == Some("deciduous") {
        minutes *= 1.3;
    }

    let time_text = if minutes < 60.0 {
        format!("~{} minutes per year", minutes.round() as i32)
    } else if minutes < 180.0 {
        format!("~{:.1} hours per year", minutes / 60.0)
    } else {
        format!("~{} hours per year", (minutes / 60.0).round() as i32)
    };

    format!("⏰ **Time Commitment**: {}", time_text)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_maintenance_high_c() {
        let data = MaintenanceData {
            csr_c: Some(0.7),
            csr_s: Some(0.2),
            csr_r: Some(0.1),
            growth_form: Some("tree".to_string()),
            height_m: Some(10.0),
            leaf_phenology: Some("deciduous".to_string()),
            decomposition_rating: None,
            decomposition_confidence: None,
        };
        let profile = generate_maintenance_profile(&data);
        assert!(profile.contains("HIGH") || profile.contains("Fast"));
    }

    #[test]
    fn test_maintenance_low_s() {
        let data = MaintenanceData {
            csr_c: Some(0.2),
            csr_s: Some(0.7),
            csr_r: Some(0.1),
            growth_form: Some("shrub".to_string()),
            height_m: Some(2.0),
            leaf_phenology: Some("evergreen".to_string()),
            decomposition_rating: None,
            decomposition_confidence: None,
        };
        let profile = generate_maintenance_profile(&data);
        assert!(profile.contains("LOW") || profile.contains("Slow"));
    }
}
